neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,443 @@
+        Self {
+            search_table,
+            header_size: Self::header_size(remote_ip),
+            mtu: search_table[probe_index],
+            probe_index,
+            probe_count: 0,
+            probe_state: Probe::NotNeeded,
+            loss_counts: vec![0; search_table.len()],
+            raise_timer: None,
+        }
+    }
+
+    /// Checks whether the PMTUD raise timer should be fired, and does so if needed.
+    pub fn maybe_fire_pmtud_raise_timer(&mut self, now: Instant) {
+        if let Some(raise_timer) = self.raise_timer {
+            if self.probe_state == Probe::NotNeeded && now >= raise_timer {
+                qdebug!("PMTUD raise timer fired");
+                self.raise_timer = None;
+                self.start_pmtud();
+            }
+        }
+    }
+
+    /// Returns the current Packetization Layer Path MTU, i.e., the maximum UDP payload that can be
+    /// sent. During probing, this may be smaller than the actual path MTU.
+    #[must_use]
+    pub fn plpmtu(&self) -> usize {
+        self.mtu - self.header_size
+    }
+
+    /// Returns true if a PMTUD probe should be sent.
+    #[must_use]
+    pub fn needs_probe(&self) -> bool {
+        self.probe_state == Probe::Needed
+    }
+
+    /// Returns the size of the current PMTUD probe.
+    fn probe_size(&self) -> usize {
+        self.search_table[self.probe_index] - self.header_size
+    }
+
+    /// Sends a PMTUD probe.
+    pub fn send_probe(
+        &mut self,
+        builder: &mut PacketBuilder,
+        stats: &mut Stats,
+        aead_expansion: usize,
+    ) {
+        builder.set_limit(self.probe_size() - aead_expansion);
+        // The packet may include ACK-eliciting data already, but rather than check for that, it
+        // seems OK to burn one byte here to simply include a PING.
+        builder.encode_varint(FRAME_TYPE_PING);
+        builder.enable_padding(true);
+        stats.frame_tx.ping += 1;
+        stats.frame_tx.all += 1;
+        stats.pmtud_tx += 1;
+        self.probe_count += 1;
+        self.probe_state = Probe::Sent;
+        qdebug!(
+            "Sending PMTUD probe of size {}, count {}",
+            self.search_table[self.probe_index],
+            self.probe_count
+        );
+    }
+
+    /// Returns true if the packet is a PMTUD probe.
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent && p.len() == self.probe_size()
+    }
+
+    /// Count the PMTUD probes included in `pkts`.
+    fn count_pmtud_probes(&self, pkts: &[SentPacket]) -> usize {
+        pkts.iter().filter(|p| self.is_pmtud_probe(p)).count()
+    }
+
+    /// Checks whether a PMTUD probe has been acknowledged, and if so, updates the PMTUD state.
+    /// May also initiate a new probe process for a larger MTU.
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        let acked = self.count_pmtud_probes(acked_pkts);
+        if acked == 0 {
```

Comment:
Do I understand correctly that `acked` could be `>1` in the case where a probe from a previous iteration is acked?

Reply:
Or if the ACK got delayed sufficiently by the sender so that it includes multiple probes.

Reply:
That's right.  We can declare something lost, probe again and then find that we have two acknowledgments to handle.  That shouldn't change anything here.